File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/clustering/partitional/Partitional.java
package cujae.inf.ic.om.heuristic.assignment.clustering.partitional;

import java.util.ArrayList;
import java.util.Random;

import cujae.inf.ic.om.factory.DistanceType;

import cujae.inf.ic.om.heuristic.assignment.clustering.Clustering;
import cujae.inf.ic.om.heuristic.assignment.clustering.SeedType;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;

import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

public abstract class Partitional extends Clustering {
	
	//todos
	protected ArrayList<Depot> createCentroids(ArrayList<Integer> idElements){
		ArrayList<Depot> centroids = new ArrayList<Depot>();
		
		for(int i = 0; i < idElements.size(); i++)
		{
			Depot centroid = new Depot();
			
			centroid.setIDDepot(idElements.get(i));
			
			Location location = new Location();
			location.setAxisX(Problem.getProblem().getCustomerByIDCustomer(idElements.get(i).intValue()).getLocationCustomer().getAxisX());
			location.setAxisY(Problem.getProblem().getCustomerByIDCustomer(idElements.get(i).intValue()).getLocationCustomer().getAxisY());
			centroid.setLocationDepot(location);
			
			centroids.add(centroid);
		}
		
		return centroids;
	}
	
	//todos
	protected void cleanClusters(ArrayList<Cluster> clusters) {
		for (Cluster c : clusters) 
			c.cleanCluster(); 
		
		System.out.println("--------------------------------------------------");
		System.out.println("LISTA DE CLUSTERS");
		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println("ID CLUSTER: " + clusters.get(i).getIDCluster());
			System.out.println("DEMANDA DEL CLUSTER: " + clusters.get(i).getRequestCluster());
			System.out.println("ELEMENTOS DEL CLUSTER: " + clusters.get(i).getItemsOfCluster());
		}
		System.out.println("--------------------------------------------------");
	}
	
	//todos
	protected ArrayList<Cluster> stepAssignment(ArrayList<Cluster> clusters, ArrayList<Customer> customerToAssign, NumericMatrix costMatrix){
		int idDepot = -1;
		int posDepot = -1;				
		double capacityDepot = 0.0;			

		int idCustomer = -1;
		int posCustomer = -1;
		double requestCustomer = 0.0;

		int posCluster = -1;
		double requestCluster = 0.0;

		RowCol rcBestAll = new RowCol();
		
		ArrayList<Customer> listCustomers = new ArrayList<Customer>(customerToAssign);
		int totalCustomers = customerToAssign.size();
		int totalDepots = clusters.size();
		
		System.out.println("--------------------------------------------------------------------");
		System.out.println("PROCESO DE ASIGNACIÓN");
		
		while((!customerToAssign.isEmpty()) && (!costMatrix.fullMatrix(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1), Double.POSITIVE_INFINITY))) 
		{
			rcBestAll = costMatrix.indexLowerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));

			posCustomer = rcBestAll.getCol();
			idCustomer = listCustomers.get(posCustomer).getIDCustomer();				
			requestCustomer = listCustomers.get(posCustomer).getRequestCustomer();
		
			System.out.println("-----------------------------------------------------------");
			System.out.println("BestAllCol: " + rcBestAll.getCol());
			System.out.println("BestAllRow: " + rcBestAll.getRow());
			
			System.out.println("ID CLIENTE SELECCIONADO: " + idCustomer);
			System.out.println("POSICIÓN DEL CLIENTE SELECCIONADO: " + posCustomer);
			System.out.println("DEMANDA DEL CLIENTE SELECCIONADO: " + requestCustomer);
			
			posDepot = (rcBestAll.getRow() - totalCustomers); 
			idDepot = Problem.getProblem().getDepots().get(posDepot).getIDDepot();
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(idDepot);

			System.out.println("ID DEPOSITO SELECCIONADO: " + idDepot);
			System.out.println("POSICIÓN DEL DEPOSITO SELECCIONADO: " + posDepot);
			System.out.println("CAPACIDAD TOTAL DEL DEPOSITO SELECCIONADO: " + capacityDepot);				
		
			posCluster = findCluster(idDepot, clusters);	
			
			System.out.println("POSICION DEL CLUSTER: " + posCluster);
			
			if(posCluster != -1)
			{
				requestCluster = clusters.get(posCluster).getRequestCluster();

				System.out.println("DEMANDA DEL CLUSTER: " + requestCluster);

				if(capacityDepot >= (requestCluster + requestCustomer)) 
				{
					requestCluster += requestCustomer;

					clusters.get(posCluster).setRequestCluster(requestCluster);
					clusters.get(posCluster).getItemsOfCluster().add(idCustomer);		

					System.out.println("DEMANDA DEL CLUSTER ACTUALIZADA: " + requestCluster);
					System.out.println("ELEMENTOS DEL CLUSTER: " + clusters.get(posCluster).getItemsOfCluster());
					
					costMatrix.fillValue(totalCustomers, posCustomer, (totalCustomers + totalDepots - 1), posCustomer, Double.POSITIVE_INFINITY);
					customerToAssign.remove(Problem.getProblem().findPosCustomer(customerToAssign, idCustomer));
					
					System.out.println("CANTIDAD DE CLIENTES SIN ASIGNAR: " + customerToAssign.size());
				}
				else
					costMatrix.setItem(rcBestAll.getRow(), posCustomer, Double.POSITIVE_INFINITY);
				
				if(isFullDepot(customerToAssign, requestCluster, capacityDepot))
				{
					System.out.println("DEPOSITO LLENO");
					
					costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.POSITIVE_INFINITY);
				}
			}
		}
		
		System.out.println("--------------------------------------------------");
		System.out.println("LISTA DE CLUSTERS");
		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println("ID CLUSTER: " + clusters.get(i).getIDCluster());
			System.out.println("DEMANDA DEL CLUSTER: " + clusters.get(i).getRequestCluster());
			System.out.println("CANTIDAD DE ELEMENTOS EN EL CLUSTER: " + clusters.get(i).getItemsOfCluster().size());
			System.out.println("ELEMENTOS DEL CLUSTER: " + clusters.get(i).getItemsOfCluster());
		}
		System.out.println("--------------------------------------------------");
		
		return clusters;
	}
	
	//todos
	protected void updateCustomerToAssign(ArrayList<Customer> customerToAssign, ArrayList<Integer> idElements){
		
		for(int i = 0; i < idElements.size(); i++)
		{
			boolean found = false;
			int j = 0;
			
			while ((!found) && j < customerToAssign.size())
			{
				if(customerToAssign.get(j).getIDCustomer() == idElements.get(i).intValue())
				{
					found = true;
					customerToAssign.remove(j);
				}
				else
					j++;
			}
		}
		
		System.out.println("CLIENTES A ASIGNAR");
		
		for(int i = 0; i < customerToAssign.size(); i++)
		{
			System.out.println("--------------------------------------------------");
			System.out.println("ID CLIENTE: " + customerToAssign.get(i).getIDCustomer());
			System.out.println("X: " + customerToAssign.get(i).getLocationCustomer().getAxisX());
			System.out.println("Y: " + customerToAssign.get(i).getLocationCustomer().getAxisY());
			System.out.println("DEMANDA: " + customerToAssign.get(i).getRequestCustomer());
		}
	}
	
	//k-means y pam
	protected ArrayList<Integer> generateElements(SeedType seedType, DistanceType distanceType){
		ArrayList<Integer> idElements = new ArrayList<Integer>();
		
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		int totalDepots = Problem.getProblem().getTotalDepots();
		int counter = totalDepots;
		
		NumericMatrix costMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		
		RowCol rcBestAll = new RowCol();
		int idElement = -1 ;
		
		switch(seedType.ordinal()) 
		{
			case 0:
			{
				for(int i = 0; i < counter; i++)
					idElements.add(-1);
				
				System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);
			
				while(counter > 0)
				{
					rcBestAll = costMatrix.indexBiggerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));
					
					System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
					System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
					System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
					
					idElement = Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer();	
					idElements.set((rcBestAll.getRow() - totalCustomers), idElement);	
					
					System.out.println("ELEMENTO: " + idElement); 
					System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);
					
					costMatrix.fillValue(totalCustomers, rcBestAll.getCol(), (totalCustomers + totalDepots - 1), rcBestAll.getCol(), Double.NEGATIVE_INFINITY);
					costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.NEGATIVE_INFINITY);
					counter--;
				}
				
				break;
			}
			case 1:
			{
				for(int i = 0; i < counter; i++)
					idElements.add(-1);
				
				System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);
				
				while(counter > 0)
				{
					rcBestAll = costMatrix.indexLowerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));
					
					System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
					System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
					System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
					
					
					idElement = Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer();	
					idElements.set((rcBestAll.getRow() - totalCustomers), idElement);	
					
					System.out.println("ELEMENTO: " + idElement); 
					System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);
					
					costMatrix.fillValue(totalCustomers, rcBestAll.getCol(), (totalCustomers + totalDepots - 1), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
					costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.POSITIVE_INFINITY);
					counter--;
				}
				
				break;
			}
			case 2:
			{
				Random rdm = new Random();
				
				while(counter > 0)
				{
					idElement = rdm.nextInt(totalCustomers); // DUDA + 1
					idElements.add(Problem.getProblem().getCustomers().get(idElement).getIDCustomer());
					
					System.out.println("ELEMENTO: " + idElement); 
					System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);
					
					counter--;
				}
				
				break;
			}
		}

		System.out.println("--------------------------------------------------");
		System.out.println("CENTROIDES/MEDOIDES INICIALES");
		System.out.println(idElements);
		System.out.println("--------------------------------------------------");
		
		return idElements;
	}
		
	//----------------------------------------------------------------------------------------------//
	
	/*
	protected ArrayList<Integer> generateElementsXXX(DistanceType distanceType){
		
		ArrayList<Integer> idElements = new ArrayList<Integer>();
		
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		int totalDepots = Problem.getProblem().getTotalDepots();
		int counter = totalDepots;

		NumericMatrix costMatrix = null;
		RowCol rcBestAll = new RowCol();
		int posElement = -1 ;
		
		switch(distanceType.ordinal())
		{
			case 0: case 1: case 2: case 3:
			{
				try {
					costMatrix = Problem.getProblem().fillCostMatrixXXX(Problem.getProblem().getCustomers(), Problem.getProblem().getDepots(), distanceType);
				} catch (IllegalArgumentException | SecurityException | ClassNotFoundException | InstantiationException
						| IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
					e.printStackTrace();
				}
	
				break;
			}
			case 4:
			{
				//costMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix()); // NO FUNcIONA
	
				break;
			}
		}
		
		
		Random rdm = new Random();
		posElement = rdm.nextInt(totalCustomers);
		idElements.add(Problem.getProblem().getCustomers().get(posElement).getIDCustomer());
		
		System.out.println("POS_ELEMENTO: " + posElement); 
		System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);
		
		while(counter > 1)
		{
			rcBestAll = costMatrix.indexBiggerValue(posElement, 0, posElement, (totalCustomers - 1));
			
			System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
			System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
			System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
	
			idElements.add(Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer());	
			posElement = rcBestAll.getCol(); 
					
			System.out.println("POS_ELEMENTO: " + posElement); 
			System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);

			costMatrix.fillValue(0, posElement, (totalCustomers - 1), posElement, Double.NEGATIVE_INFINITY); // no es necesario
			counter--;
		}
		
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);
	
		return sortedElements(idElements, distanceType);
	}
	 */
	/*
	protected ArrayList<Integer> generateElementsXX(DistanceType distanceType){
		
		ArrayList<Integer> listCentroids = new ArrayList<Integer>();
		ArrayList<Customer> listCustomers = new ArrayList<Customer>(Problem.getProblem().getCustomers());

		int totalCustomers = listCustomers.size();
		int counter = Problem.getProblem().getTotalDepots();
		int idElement = -1;

		Random rdm = new Random();
		int posCentroid = rdm.nextInt(totalCustomers);
		
		idElement = Problem.getProblem().getCustomers().get(posCentroid).getIDCustomer();
		listCentroids.add(idElement);
		listCustomers.remove(posCentroid);

		counter--;
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + listCentroids);

		NumericMatrix costMatrix = null;
	
		switch(distanceType.ordinal())
		{
			case 0: case 1: case 2: case 3:
			{
				try {
					costMatrix = Problem.getProblem().fillCostMatrixXXX(Problem.getProblem().getCustomers(), Problem.getProblem().getDepots(), distanceType);
				} catch (IllegalArgumentException | SecurityException | ClassNotFoundException | InstantiationException
						| IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
					e.printStackTrace();
				}
				break;
			}
			case 4:
			{
				costMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
	
				break;
			}
		}

		RowCol rcBestAll = new RowCol();
		rcBestAll = costMatrix.indexBiggerValue(posCentroid, 0, posCentroid, (totalCustomers - 1));
		
		idElement = Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer();
		listCentroids.add(idElement);
		listCustomers.remove(Problem.getProblem().getPosElement(idElement, listCustomers));
		
		counter--;
		
		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + listCentroids);
		
		ArrayList<Double> maxDistances = new ArrayList<Double>();
		int posCustomer = -1;

		while(counter > 0)
		{
			double maxDistance = -1;
			double currentDistance = -1;
			
			for(int i = 0; i < listCustomers.size(); i++)
			{
				posCustomer = Problem.getProblem().getPosElement(listCustomers.get(i).getIDCustomer());
				
				for(int j = 0; j < listCentroids.size(); j++)
				{
					posCentroid = Problem.getProblem().getPosElement(listCentroids.get(j).intValue());
					currentDistance = costMatrix.getItem(posCentroid, posCustomer);

					if(j == 0)
						maxDistance = currentDistance;
					else
						if(maxDistance < currentDistance)
							maxDistance = currentDistance;
				}
				
				maxDistances.add(maxDistance);
			}

			double minDistance = maxDistances.get(0);
			posCentroid = 0;
			
			for(int k = 1; k < maxDistances.size(); k++)
			{
				if(maxDistances.get(k) < minDistance)
				{
					minDistance = maxDistances.get(k);
					posCentroid = k;
				}
			}

			idElement = listCustomers.get(posCentroid).getIDCustomer();
			listCentroids.add(idElement);

			listCustomers.remove(posCentroid);
			counter--;

			System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + listCentroids);
		}

		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + listCentroids);

		return sortedElements(listCentroids, distanceType);
		}
	 */
	/*
	protected ArrayList<Integer> generateElements(ArrayList<Integer> pCustomers, int sampsize, DistanceType distanceType){

		ArrayList<Integer> idElements = new ArrayList<Integer>();

		int totalCustomers = Problem.getProblem().getTotalCustomers();
		int totalDepots = Problem.getProblem().getTotalDepots();
		int counter = totalDepots;

		//Generar matriz de costo con las distancias
		NumericMatrix costMatrix = null;
		RowCol rcBestAll = new RowCol();
		int idElement = -1;

		switch(distanceType.ordinal())
		{
		case 0: case 1: case 2: case 3:
		{
			/*	try {
						costMatrix = InfoProblem.getProblem().fillCostMatrix(generatePartitions(sampsize), InfoProblem.getProblem().getDepots(), distanceType);
					} 
					catch (IllegalArgumentException | SecurityException | ClassNotFoundException | InstantiationException 
							| IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
						e.printStackTrace();
					}*/
/*
			break;
		}

		case 4:
		{
			costMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
			break;
		}
		}

		//Asignar elementos a los medoides
		for(int i = 0; i < counter; i++)
			idElements.add(-1);

		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);

		while(counter > 0)
		{
			rcBestAll = costMatrix.indexLowerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));

			System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
			System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
			System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));


			idElement = Problem.getProblem().getCustomers().get(rcBestAll.getCol()).getIDCustomer();	
			idElements.set((rcBestAll.getRow() - totalCustomers), idElement);	

			System.out.println("ELEMENTO: " + idElement); 
			System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);

			costMatrix.fillValue(totalCustomers, rcBestAll.getCol(), (totalCustomers + totalDepots - 1), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
			costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.POSITIVE_INFINITY);
			counter--;
		}

		System.out.println("--------------------------------------------------");
		System.out.println("CENTROIDES/MEDOIDES INICIALES");
		System.out.println(idElements);
		System.out.println("--------------------------------------------------");

		return idElements;
	}
	*/
	/*
	protected double calculateCost(ArrayList<Cluster> clusters, NumericMatrix costMatrix ) {
		double cost = 0.0;

		int idCustomer =  -1;
		int posCustomer = -1;
		int posDepot = -1;
		ArrayList<Integer> listIDCustomers;

		for(int i = 0; i < clusters.size(); i++) 
		{
			listIDCustomers = clusters.get(i).getItemsOfCluster();
			posDepot = Problem.getProblem().getPosElement(clusters.get(i).getIDCluster());
			
			System.out.println("CLIENTES: " + listIDCustomers);
			System.out.println("POSICIÓN DEPOSITO: " + posDepot);
			
			for(int j = 0; j < listIDCustomers.size(); j++) 
			{	
				idCustomer =  listIDCustomers.get(j);
				posCustomer = Problem.getProblem().getPosElement(idCustomer);

				cost += costMatrix.getItem(posDepot, posCustomer);	
				
				System.out.println("ID CLIENTE: " + idCustomer);
				System.out.println("POSICIÓN CLIENTE: " + posCustomer);
				System.out.println("COSTO: " + costMatrix.getItem(posDepot, posCustomer));
				System.out.println("COSTO ACUMULADOS: " + cost);	
			}
		}

		return cost;
	}
	*/
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/partitional/partitional.py
import random
import numpy as np
from typing import List
from ..seed_type import SeedType
from ..clustering import Clustering
from ....factory.interfaces.assignment_type import AssignmentType
from ....service.distance_type import DistanceType
from ....problem.input.problem import Problem
from ....problem.input.problem import Customer
from ....problem.input.problem import Depot
from ....problem.input.location import Location
from ....problem.solution.cluster import Cluster

class Partitional(Clustering):
    
    def __init__(self):
        super().__init__()
        self.seed_type = SeedType.NEAREST_DEPOT
        self.count_max_iterations = 100   # Configurable
        self.current_iteration = 0
    
    # Método que genera elementos iniciales (centroides o medoides) para los clústeres.
    def generate_elements(self) -> List[int]:
        id_elements: List[int] = []
        total_depots: int = Problem.get_problem().get_total_depots()
        total_customers = Problem.get_problem().get_total_customers()
        counter = total_depots

        print(f"LISTADO DE ELEMENTOS SELECCIONADOS: {id_elements}")

        if self.seed_type == SeedType.FARTHEST_DEPOT:
            depot = Depot()
            depot.set_id_depot(-1)
            depot.set_location_depot(self.calculate_mean_coordinate())
            list_depot = [depot]
            
            cost_matrix: np.ndarray = self.initialize_cost_matrix(
                Problem.get_problem().get_customers(), list_depot, self.distance_type)
            
            while counter > 0:
                max_value_index = np.argmax(cost_matrix)
                row = max_value_index // cost_matrix.shape[1]
                col = max_value_index % cost_matrix.shape[1]
                
                print(f"FILA SELECCIONADA: {row}")
                print(f"COLUMNA SELECCIONADA: {col}")
                print(f"VALOR SELECCIONADO: {cost_matrix[row, col]}")
                
                id_element = Problem.get_problem().get_customers()[col].get_id_customer()
                id_elements.append(id_element)
                
                print(f"ELEMENTO: {id_element}")
                print(f"LISTADO DE ELEMENTOS ACTUALIZADOS: {id_elements}")
                    
                cost_matrix[row, col] = float('-inf')
                counter -= 1   
        
        elif self.seed_type == SeedType.NEAREST_DEPOT:

            cost_matrix: np.ndarray = Problem.get_problem().get_cost_matrix()
            if cost_matrix is None or not isinstance(cost_matrix, (np.ndarray, list)) or len(cost_matrix) == 0:
                cost_matrix = self.initialize_cost_matrix(
                    Problem.get_problem().get_customers(), Problem.get_problem().get_depots(), self.distance_type
                )
            
            while counter > 0:
                min_value_index = np.argmin(cost_matrix)
                row = min_value_index // cost_matrix.shape[1]
                col = min_value_index % cost_matrix.shape[1]
            
                print(f"FILA SELECCIONADA: {row}")
                print(f"COLUMNA SELECCIONADA: {col}")
                print(f"VALOR SELECCIONADO: {cost_matrix[row, col]}")
                
                id_element = Problem.get_problem().get_customers()[col].get_id_customer()
                id_elements.append(id_element)
                
                print(f"ELEMENTO: {id_element}")
                print(f"LISTADO DE ELEMENTOS ACTUALIZADOS: {id_elements}")
                
                cost_matrix[row, col] = float('-inf')
                counter -= 1
        
        elif self.seed_type == SeedType.RANDOM_DEPOT:
            random.seed()
            
            while counter > 0:
                id_element = random.randint(0, total_customers - 1)
                id_elements.append(Problem.get_problem().get_customers()[id_element].get_id_customer())
                
                print(f"ELEMENTO: {id_element}")
                print(f"LISTADO DE ELEMENTOS ACTUALIZADOS: {id_elements}")
                
                counter -= 1
            
        print(f"--------------------------------------------------")
        print(f"CENTROIDES/MEDOIDES INICIALES")
        print(f"{id_elements}")
        print(f"--------------------------------------------------")
            
        return id_elements 
    
    # Método que ordena los elementos por proximidad a los depósitos según el tipo de distancia, 
    # utilizando una matriz de costos para determinar el orden.
    def sorted_elements(self, id_elements: List[int], distance_type: DistanceType) -> List[int]:
        total_depots = Problem.get_problem().get_total_depots()
        j = 0
        
        sorted_elements: List[int] = [-1] * len(id_elements)
        customers: List[Customer] = []
        
        for element_id in id_elements:
            customers.append(Problem.get_problem().get_customer_by_id_customer(element_id))

        cost_matrix: np.ndarray = self.initialize_cost_matrix(
            customers, Problem.get_problem().get_depots(), distance_type)
                
        while j < len(id_elements):
            min_index = np.argmin(cost_matrix)
            row = min_index // cost_matrix.shape[1]
            col = min_index % cost_matrix.shape[1]
            
            print(f"ROW SELECCIONADA: {row}")
            print(f"COL SELECCIONADA: {col}")
            print(f"VALOR SELECCIONADO: {cost_matrix[row, col]}")

            cost_matrix[0:len(id_elements), col] = float('inf')
            cost_matrix[row, 0:len(id_elements) + total_depots - 1] = float('inf')

            sorted_elements[col - len(id_elements)] = id_elements[row]

            print(f"LISTADO DE ELEMENTOS SELECCIONADOS ORDENADOS ACTUALIZADA: {sorted_elements}")
            j += 1
            
        print(f"LISTADO DE ELEMENTOS SELECCIONADOS ORDENADOS: {sorted_elements}")

        return sorted_elements
    
    # Método que calcula y retorna la coordenada promedio de todas las ubicaciones de los clientes.
    def calculate_mean_coordinate(self) -> Location:
        axis_x: float = 0.0
        axis_y: float = 0.0

        list_coordinates_customers: List[Location] = list(Problem.get_problem().get_list_coordinates_customers())

        for customer in list_coordinates_customers:
            axis_x += customer.get_axis_x()
            axis_y += customer.get_axis_y()

        axis_x /= len(list_coordinates_customers)
        axis_y /= len(list_coordinates_customers)

        mean_location = Location(axis_x, axis_y)

        return mean_location
    
    # Método de asignación de clientes a clusters según los depósitos, basado en la matriz de costos, 
    # la demanda de los clientes y la capacidad de los depósitos.
    def step_assignment(self, list_clusters: List[Cluster], list_centroids: List[Depot]) -> List[Cluster]:
        total_customers = len(self.list_customers_to_assign)
        
        print("--------------------------------------------------------------------")
        print(f"PROCESO DE ASIGNACIÓN")
        
        while self.list_customers_to_assign:
            cost_matrix: np.ndarray = self.initialize_cost_matrix(
                self.list_customers_to_assign, 
                list_centroids,
                self.distance_type
            )
            
            min_index = np.argmin(cost_matrix)
            row_best = min_index // cost_matrix.shape[1]
            col_best = min_index % cost_matrix.shape[1]
            
            selected_customer = self.list_customers_to_assign[col_best]
            id_customer = selected_customer.get_id_customer()
            request_customer = selected_customer.get_request_customer()
            
            print("-----------------------------------------------------------")
            print(f"ID CLIENTE SELECCIONADO: {id_customer}")
            print(f"POSICIÓN DEL CLIENTE SELECCIONADO: {col_best}")
            print(f"DEMANDA DEL CLIENTE SELECCIONADO: {request_customer}")
            
            selected_depot: Depot = Problem.get_problem().get_depots()[row_best]
            id_depot = selected_depot.get_id_depot()
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(id_depot)
            
            print(f"ID DEPOSITO SELECCIONADO: {id_depot}")
            print(f"POSICIÓN DEL DEPOSITO SELECCIONADO: {row_best}")
            print(f"CAPACIDAD TOTAL DEL DEPOSITO SELECCIONADO: {capacity_depot}")

            pos_cluster = self.find_cluster(id_depot, list_clusters)
            
            print(f"POSICION DEL CLUSTER: {pos_cluster}")

            if pos_cluster != -1:
                cluster: Cluster = self.list_clusters[pos_cluster]
                
                print(f"DEMANDA DEL CLIENTE: {request_customer}")
                print(f"CAPACIDAD DEL DEPÓSITO: {capacity_depot}")

                new_demand = cluster.get_request_cluster() + request_customer
    
                print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")

                if capacity_depot >= new_demand:
                    cluster.set_request_cluster(new_demand)
                    items_of_cluster: List[int] = cluster.get_items_of_cluster()
                    items_of_cluster.append(id_customer)
                    cluster.set_items_of_cluster(items_of_cluster)

                    print(f"DEMANDA DEL CLUSTER ACTUALIZADA: {new_demand}")
                    print(f"ELEMENTOS DEL CLUSTER: {items_of_cluster}")

                    self.list_customers_to_assign.remove(selected_customer)

                    print(f"CANTIDAD DE CLIENTES SIN ASIGNAR: {len(self.list_customers_to_assign)}")
                else:
                    cost_matrix[row_best, col_best] = float('inf')

                if self.is_full_depot(self.list_customers_to_assign, cluster.get_request_cluster(), capacity_depot):
                    print("DEPOSITO LLENO")

                    cost_matrix[row_best, 0:total_customers] = float('inf')

        print("--------------------------------------------------")
        print("LISTA DE CLUSTERS")

        for cluster in list_clusters:
            print(f"ID CLUSTER: {cluster.id_cluster}")
            print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")
            print(f"CANTIDAD DE ELEMENTOS EN EL CLUSTER: {len(cluster.get_items_of_cluster())}")
            print(f"ELEMENTOS DEL CLUSTER: {cluster.get_items_of_cluster()}")
        print("--------------------------------------------------")

        return list_clusters
         
    # Método que crea una lista de centroides a partir de los identificadores de elementos, 
    # asignando a cada uno su ubicación correspondiente según la posición de los clientes.
    def create_centroids(self) -> List[Depot]:
        centroids: List[Depot] = []

        for i in range(len(self.list_id_elements)):
            centroid = Depot()
            customer = Problem.get_problem().get_customer_by_id_customer(self.list_id_elements[i])

            if customer is not None:  # Verificar si el cliente existe
                centroid.set_id_depot(self.list_id_elements[i])

                location = Location()
                location.set_axis_x(customer.get_location_customer().get_axis_x())
                location.set_axis_y(customer.get_location_customer().get_axis_y())
                centroid.set_location_depot(location)

                centroids.append(centroid)
            else:
                print(f"Cliente con ID {self.list_id_elements[i]} no encontrado.")

        return centroids
        
    # Método para actualizar la lista de clientes por asignar, eliminando aquellos cuyos IDs 
    # coincidan con los elementos especificados en una lista de IDs.
    def update_customer_to_assign(self):
        
        for id_element in self.list_id_elements:
            found: bool = False
            i: int = 0
            while not found and i < len(self.list_customers_to_assign):
                if self.list_customers_to_assign[i].get_id_customer() == id_element:
                    found = True
                    del self.list_customers_to_assign[i]
                else:
                    i += 1
                    
        print("CLIENTES A ASIGNAR")
        for customer in self.list_customers_to_assign:
            print(f"--------------------------------------------------")
            print(f"ID CLIENTE: {customer.get_id_customer()}")
            print(f"X: {customer.get_location_customer().get_axis_x()}")
            print(f"Y: {customer.get_location_customer().get_axis_y()}")
            print(f"DEMANDA: {customer.get_request_customer()}")
            
    # Método que limpia los clusters eliminando sus elementos y muestra información detallada 
    # de cada cluster.
    def clean_clusters(self, clusters: List[Cluster]):
        for cluster in clusters:
            cluster.clean_cluster()
        
        print("--------------------------------------------------")
        print("LISTA DE CLUSTERS")
        for i, cluster in enumerate(clusters):
            print(f"ID CLUSTER: {cluster.get_id_cluster()}")
            print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")
            print(f"ELEMENTOS DEL CLUSTER: {cluster.get_items_of_cluster()}")
        print("--------------------------------------------------")
        